File: p04ab.hs
-- Advent of Code 2025 - Day 4 part One and Two
-- Solutions in Haskell
-- (Ter leering ende vermaeck...)
--
--  Part one: The number of rolls of paper that can be accessed by a forklift is: 1370
--  Part two: The number of rolls of paper that can be accessed by a forklift is: 8437
--
-- (cl) by Arno Jacobs, 2025-12-04

module AoC2025d04ab where

data Part = PartOne | PartTwo deriving Eq

filename :: String
filename = "data/inputDay04_2025.txt"

type Element        = Char
type PaperRollGrid  = [[Element]]
type Counter        = Int
type CountGrid      = [[Counter]]
type Position       = (Int,Int)

paperRoll :: Element
paperRoll = '@'

cleanSpace :: Element
cleanSpace = '.'

countAdjacentRolls :: PaperRollGrid -> CountGrid
countAdjacentRolls grid = [[ countAdjacentRolls' grid (x,y) (maxX,maxY) | x <- [0..maxX]] | y <- [0..maxY]] 
    where
        maxX = length (grid !! 0) - 1
        maxY = length grid - 1
    --  countAdjacentRolls' :: PaperRollGrid -> Position -> Position -> Counter
        countAdjacentRolls' grid (x,y) (maxX,maxY)
            | grid !! y !! x /= paperRoll   = 0
            | otherwise                     =
                sum [ countAdjacentRolls'' grid (x+dx,y+dy) (maxX,maxY) | dx <- [-1..1], dy <- [-1..1] ]
            where            
            --  countAdjacentRolls'' :: PaperRollGrid -> Position -> Position -> Counter
                countAdjacentRolls'' grid (dx,dy) (maxX,maxY)
                    |   dx < 0      ||  dy < 0 
                    ||  dx > maxX   ||  dy > maxY       = 0
                    |   grid !! dy !! dx == paperRoll   = 1 
                    |   otherwise                       = 0

removePaperRolls :: CountGrid -> PaperRollGrid
removePaperRolls countGrid = [[ removePaperRoll (countGrid !! y !! x) | x <- [0..maxX]] | y <- [0..maxY]] 
    where
        maxX = length (countGrid !! 0) - 1
        maxY = length countGrid - 1
    --  removePaperRoll :: Counter -> Element
        removePaperRoll prc | prc < 5   = cleanSpace
                            | otherwise = paperRoll

workGrid :: Part -> Counter -> PaperRollGrid -> Counter
workGrid part cnt grid  | part == PartOne   = canBeMoved
                        | canBeMoved == 0   = cnt
                        | otherwise         = workGrid part (canBeMoved + cnt) nextGrid
    where
        countGrid   = countAdjacentRolls grid
        canBeMoved  = (length . filter canAccess . concat) countGrid
        nextGrid    = removePaperRolls countGrid
    --  canAccess :: Counter -> Bool
        canAccess prc = prc > 0 && prc < 5  
        --  < 5 because the code is also counting the paper roll that can be accessed itself


main :: IO ()
main = do   putStrLn "Advent of Code 2025 - day 4  (Haskell)"
            day4 <- lines <$> readFile filename
            putStr "Part one: The number of rolls of paper that can be accessed by a forklift is: "
            print $ workGrid PartOne 0 day4
            putStr "Part two: The number of rolls of paper that can be accessed by a forklift is: "
            print $ workGrid PartTwo 0 day4
            putStrLn "0K.\n"

-- End of code

File: p04ab.clj
;;;;    Advent of Code 2025 - Day 4 part One and Two
;;;;    https://adventofcode.com/2025/day/4
;;;;
;;;;    Solutions in Clojure
;;;;    (Ter leering ende vermaeck...)
;;;;
;;;;    Part one: The number of rolls of paper that can be removed by a forklift is: 1370
;;;;    Part two: The number of rolls of paper that can be removed by a forklift is: 8437
;;;;
;;;;    (cl) by Arno Jacobs, 2025-12-24
;;;;    

(ns p04ab
  (:require [clojure.string :as str]))

;;; Some helper defines
(def paper-roll \@)
(def clean-space \.)
(def part1 1)
(def part2 2)

;;;; Read the data-set
(defn get-lines
  [file]
  (->> file
       slurp
       (str/split-lines)))

(def grid (get-lines "./data/inputDay04_2025.txt"))

;;;; helper - pretty print grid 
(defn pretty-grid
  [grid]
  (doseq [line grid]
    (println line)))

;;;; Helper for counting neighbouring rolls
(defn add-roll
  [grid x y max-x max-y]
  (if (and (>= x 0) (>= y 0) (< x max-x) (< y max-y))
    (if (= paper-roll (nth (nth grid y) x))
      1
      0)
    0))

;;; take one paper roll if there are less than 4 neighbouring paper rolls
(defn take-paper-roll
  [grid x y max-x max-y]
  (if (= paper-roll (nth (nth grid y) x))
    (if (< (+ (add-roll grid (- x 1) (- y 1) max-x max-y) 
              (add-roll grid (- x 1)    y    max-x max-y) 
              (add-roll grid (- x 1) (+ y 1) max-x max-y) 
              (add-roll grid    x    (- y 1) max-x max-y) 
              (add-roll grid    x    (+ y 1) max-x max-y) 
              (add-roll grid (+ x 1) (- y 1) max-x max-y) 
              (add-roll grid (+ x 1)    y    max-x max-y) 
              (add-roll grid (+ x 1) (+ y 1) max-x max-y)) 4)
      clean-space
      paper-roll)
    clean-space))

;;; Work one row - column by column
(defn work-row
  [grid x y max-x max-y]
  (if (>= x max-x)
    nil
    (cons (take-paper-roll grid x y max-x max-y)
          (work-row grid (+ 1 x) y max-x max-y))))
  
;;;; Work row by row
(defn work-grid
  [grid row max-x max-y]
  (if (>= row max-y)
     nil
     (cons (work-row grid 0 row max-x max-y)
           (work-grid grid (+ 1 row) max-x max-y))))
  
(defn take-paper-rolls
  [grid]
  (let [max-x (count (first grid))
        max-y (count grid)] 
    (work-grid grid 0 max-x max-y)))

;;; Count the remaining number of paper rolls on the grid
(defn count-paper-rolls
  [grid]
  (count (filter #(= paper-roll %) (str/join grid))))

(defn count-taken-paper-rolls
  [part grid]
  (let [current-rolls         (count-paper-rolls grid)
        next-rolls            (take-paper-rolls grid)
        next-number-of-rolls  (count-paper-rolls next-rolls)
        rolls-taken           (- current-rolls next-number-of-rolls)] 
    (if (= part part1) 
      rolls-taken
      (if (= 0 rolls-taken) 
        0 
        (+ rolls-taken (count-taken-paper-rolls part next-rolls))))))

;;; The 'main' program - - -
(defn program []
  (println "Advent of Code 2025 - day 4  (Clojure)")
  (print   "Part one: The number of rolls of paper that can be removed by a forklift is: ")
  (println (count-taken-paper-rolls part1 grid))
  (print   "Part two: The number of rolls of paper that can be removed by a forklift is: ")
  (println (count-taken-paper-rolls part2 grid))
  (println "0K.\n"))

;; Run in terminal via: clojure -M p04ab.clj
(program)



